ostd/src/mm/kspace/kva.rs

```diff
@@ -76,138 +88,341 @@ impl VirtAddrAllocator {
             }
         }
     }
 }
 
-impl Drop for Kva {
+impl<M: AllocatorSelector + 'static> Drop for Kva<M> {
     fn drop(&mut self) {
         // 1. unmap all mapped pages.
         let page_table = KERNEL_PAGE_TABLE.get().unwrap();
         let range = self.start()..self.end();
         let mut cursor = page_table.cursor_mut(&range).unwrap();
-        unsafe {
-            cursor.unmap(range.len());
+        let flusher = TlbFlusher::new(CpuSet::new_full(), disable_preempt());
+        let tlb_prefer_flush_all = self.end() - self.start() > FLUSH_ALL_RANGE_THRESHOLD;
+
+        loop {
+            let result = unsafe { cursor.take_next(self.end() - cursor.virt_addr()) };
+            match result {
+                PageTableItem::Mapped { va, page, .. } => match TypeId::of::<M>() {
+                    id if id == TypeId::of::<Tracked>() => {
+                        if !flusher.need_remote_flush() && tlb_prefer_flush_all {
+                            // Only on single-CPU cases we can drop the page immediately before flushing.
+                            drop(page);
+                            continue;
+                        }
+                        flusher.issue_tlb_flush_with(TlbFlushOp::Address(va), page);
+                    }
+                    id if id == TypeId::of::<Untracked>() => {
+                        panic!("Found tracked memory mapped into untracked `Kva`");
+                    }
+                    _ => panic!("Unexpected `Kva` type"),
+                },
+                PageTableItem::MappedUntracked { va, .. } => match TypeId::of::<M>() {
+                    id if id == TypeId::of::<Untracked>() => {
+                        if !flusher.need_remote_flush() && tlb_prefer_flush_all {
+                            continue;
+                        }
+                        flusher.issue_tlb_flush(TlbFlushOp::Address(va));
+                    }
+                    id if id == TypeId::of::<Tracked>() => {
+                        panic!("Found untracked memory mapped into tracked `Kva`");
+                    }
+                    _ => panic!("Unexpected `Kva` type"),
+                },
+                PageTableItem::NotMapped { .. } => {
+                    break;
+                }
+                PageTableItem::PageTableNode { .. } => {
+                    panic!("Found page table node in `Kva`");
+                }
+            }
+        }
+
+        if !flusher.need_remote_flush() && tlb_prefer_flush_all {
+            flusher.issue_tlb_flush(TlbFlushOp::All);
         }
-        tlb_flush_addr_range(&range);
+
+        flusher.dispatch_tlb_flush();
+
         // 2. free the virtual block
-        let mut lock_guard = KVA_ALLOCATOR.lock();
-        lock_guard.deref_mut().free(range);
+        let allocator = M::select_allocator();
+        allocator.free(range);
     }
-}
\ No newline at end of file
+}
+
+impl Kva<Untracked> {
+    fn reader(&self) -> VmReader<'_, Infallible> {
```

Comment:
I don't think implementing `reader` and `writer` methods are safe in general. Whether the range covered by a `Kva<Untracked>` is determined by how `map_untracked_pages` is called. In the case of `IoMem`, the full range is mapped so creating readers and writers are ok. As you can see, the safety depends on the user of `Kva`. So I think you should move the implementation of `reader` and `writer` methods to `IoMem`.